### PR TITLE
Feature: support ISO-8601 datetimes in time travel queries

### LINKED ISSUE #223: ISO 8601 support in time travel queries

### PR DESCRIPTION
**Pull request by cap10morgan**
Closes #223

Ran into a very strange issue in the test I wrote for this (see inline comments in the `time-travel-test` ns), but o/w seems to work.

### CODE REVIEW

File: src/fluree/db/time_travel.cljc [hunk 1]
```diff
@@ -143,6 +145,29 @@
...
+  latest 't' value that is not more recent than that datetime."
+  [db datetime]
+  (go-try
+    (log/debug "datetime->t db:" (pr-str db))
+    (let [epoch-datetime (util/str->epoch-ms datetime)
+          flakes (some-> db
+                         dbproto/-rootdb
+                         (query-range/index-range :post
+                                                  > [const/$_commit:time 0]
+                                                  < [const/$_commit:time epoch-datetime])
```
Comment (bplatz) [inline src/fluree/db/time_travel.cljc:159]: Should this be `<=`?

File: src/fluree/db/time_travel.cljc [hunk 2]
```diff
@@ -143,6 +145,29 @@
...
+                         <?)]
+      (log/debug "datetime->t index-range:" (pr-str flakes)
+                 #_(keep #(when (some (fn [v] (= v const/$_commit:time)) %)
+                            %) flakes))
+      (when (empty? flakes)
+        (throw (ex-info (str "There is no data as of " datetime)
+                        {:status 400, :error :db/invalid-query})))
+      (let [flake (apply max-key flake/s flakes)]
```
Comment (bplatz) [inline src/fluree/db/time_travel.cljc:167]: The flakes will be ordered already, so you should be able to do `(peek flakes)` as opposed to applying a function over them.

**Comment by bplatz** [comment]
I've left some comments on the current approach, but I think a much more scalable approach to this problem is instead of finding all 't' values <= requested timestamp and taking the highest (last) value - which could be millions (or billions) of values if there are millions of 't's, instead do the exact opposite approach and find t values > requested timestamp then take the first result and decrement it by 1. If lazy, this would only pull the minimal set of results and we just need the first one.

(in pseudo-code):
```
(-> (range-scan db >= requested-timestamp)
    first
    (get 't' value)
    dec)
```
If there are zero results, just use the current (latest) db.

We don't have a good API to get lazy index-range results, I've been implementing a custom one on top of the tree-chan which is lazy. But even if not lazy initially, I think this will be faster most of the time, and once lazy it will definitely be very scalable regardless of how many 't' values exist.

**Comment by cap10morgan** [comment]
@bplatz Let me know what you think of [4163888](https://github.com/fluree/db/pull/273/commits/41638889ee6cf12d014ca5344ab78c102f067a6f)

File: src/fluree/db/time_travel.cljc [hunk 3]
```diff
@@ -155,18 +155,13 @@
           flakes (some-> db
                          dbproto/-rootdb
                          (query-range/index-range :post
-                                                  > [const/$_commit:time 0]
-                                                  < [const/$_commit:time epoch-datetime])
+                                                  >= [const/$_commit:time epoch-datetime]
```
Comment (bplatz) [inline src/fluree/db/time_travel.cljc:160]: I actually think this logic would require just `>`, as a perfectly equal timestamp should return the 't' value of that match.

File: src/fluree/db/time_travel.cljc [hunk 4]
```diff
@@ -155,18 +155,13 @@
...
+                                                  <  [const/$_commit:time (System/currentTimeMillis)])
```
Comment (bplatz) [inline src/fluree/db/time_travel.cljc]: CLJ-only, I think we have a CLJC version in `util`.

**Review by bplatz — APPROVED** [review]
I think once those couple comments are addressed, LGTM!

File: src/fluree/db/time_travel.cljc (hunk 4 above)
Comment (cap10morgan) [reply src/fluree/db/time_travel.cljc]: Resolved in [e122a09](https://github.com/fluree/db/pull/273/commits/e122a09cb0ccdcd5243b6dead90a70fdf5252827) but I'll see if there's a better version in `util`.

File: src/fluree/db/time_travel.cljc [hunk 5]
```diff
@@ -152,14 +152,12 @@
...
     (let [epoch-datetime (util/str->epoch-ms datetime)
-          current-time #?(:clj (System/currentTimeMillis)
-                          :cljs (js/Date.now))
-          flakes (some-> db
-                         dbproto/-rootdb
-                         (query-range/index-range :post
-                                                  >= [const/$_commit:time epoch-datetime]
-                                                  <  [const/$_commit:time current-time])
-                         <?)]
+          flakes         (some-> db
```
Comment (bplatz) [inline src/fluree/db/time_travel.cljc:155]: I don't think `some->` is doing anything here and should just be `->`

**Comment by cap10morgan** [comment]
@bplatz Do we care that this approach can no longer distinguish "you want the latest t" from "there was no data at the given time" (and just returns the latest there too)?

**Comment by bplatz** [comment]
Do you mean if you asked for 't' as of year 2100 you'd just get the latest DB.

To get the latest 't' you'd always just call `(fluree/db myledger)` - so you have the explicit way of getting the latest t always.

Maybe there is a scenario you are referring to that I'm missing.

**Comment by cap10morgan** [comment]
> Do you mean if you asked for 't' as of year 2100 you'd just get the latest DB.
> 
> To get the latest 't' you'd always just call `(fluree/db myledger)` - so you have the explicit way of getting the latest t always.
> 
> Maybe there is a scenario you are referring to that I'm missing.

No I mean that if you ask for a time before the db was created, you'd get the latest db.

**Comment by bplatz** [comment]
I think it would return the db with t=0, wouldn't it?

Assuming I'm correct we could check for the t=0 and throw with "DB did not exist".

**Comment by cap10morgan** [comment]
> I think it would return the db with t=0, wouldn't it?
> 
> Assuming I'm correct we could check for the t=0 and throw with "DB did not exist".

Yeah, never mind. Something else must have been going on in some local testing I was doing before. Currently it does seem to calculate `:t 0` and just returns `[]` for that. That seems reasonable to me, but would you prefer it throw in that case?

**Comment by bplatz** [comment]
> > I think it would return the db with t=0, wouldn't it?
> > Assuming I'm correct we could check for the t=0 and throw with "DB did not exist".
> 
> Yeah, never mind. Something else must have been going on in some local testing I was doing before. Currently it does seem to calculate `:t 0` and just returns `[]` for that. That seems reasonable to me, but would you prefer it throw in that case?

You mean any attempted query returns `[]`? If so, I think that is reasonable, but but I'd personally prefer it throw as it implies a DB exists when it doesn't. It is possible the query would return `[]` at a valid point in time as well so distinguishing between these feels like the right thing to do.

**Comment by cap10morgan** [comment]
>  but I'd personally prefer it throw as it implies a DB exists when it doesn't.

Done in [2074441](https://github.com/fluree/db/pull/273/commits/2074441c0fd992ccafbb3192b83501421647af65)

**Pull request merged**